pkg/engine/template_generator.go

```diff
@@ -1022,3 +1124,76 @@ func (t *TemplateGenerator) getTemplateFuncMap(cs *api.ContainerService) templat
 		"shellQuote": helpers.ShellQuote,
 	}
 }
+
+func (t *TemplateGenerator) GenerateTemplateV2(containerService *api.ContainerService, generatorCode string, acsengineVersion string) (templateRaw string, parametersRaw string, err error) {
+
+	armParams, _ := t.getParameterDescMap(containerService)
+	armResources := GenerateARMResources(containerService)
+	armVariables := GetKubernetesVariables(containerService)
+	armOutputs := GetKubernetesOutputs(containerService)
+
+	armTemplate := ARMTemplate{
+		Schema:         "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+		ContentVersion: "1.0.0.0",
+		Parameters:     armParams["parameters"],
+		Variables:      armVariables,
+		Resources:      armResources,
+		Outputs:        armOutputs,
+	}
+
+	templBytes, _ := json.Marshal(armTemplate)
+	templateRaw = string(templBytes)
+
+	var parametersMap paramsMap
+	if parametersMap, err = getParameters(containerService, generatorCode, acsengineVersion); err != nil {
+		return "", "", err
+	}
+
+	var parameterBytes []byte
+	if parameterBytes, err = helpers.JSONMarshal(parametersMap, false); err != nil {
+		return "", "", err
+	}
+	parametersRaw = string(parameterBytes)
+
+	return templateRaw, parametersRaw, err
+}
+
+func (t *TemplateGenerator) getParameterDescMap(containerService *api.ContainerService) (map[string]interface{}, error) {
+	var templ *template.Template
+	var paramsDescMap map[string]interface{}
+	properties := containerService.Properties
+	// save the current orchestrator version and restore it after deploying.
+	// this allows us to deploy agents on the most recent patch without updating the orchestrator version in the object
+	orchVersion := properties.OrchestratorProfile.OrchestratorVersion
+	defer func() {
+		properties.OrchestratorProfile.OrchestratorVersion = orchVersion
+	}()
+
+	templ = template.New("acs template").Funcs(t.getTemplateFuncMap(containerService))
+
+	files, baseFile := kubernetesParamFiles, kubernetesParameters
+
+	for _, file := range files {
+		bytes, e := Asset(file)
+		if e != nil {
+			err := t.Translator.Errorf("Error reading file %s, Error: %s", file, e.Error())
+			return nil, err
+		}
+		if _, err := templ.New(file).Parse(string(bytes)); err != nil {
+			return nil, err
+		}
+	}
+
+	var b bytes.Buffer
+	if err := templ.ExecuteTemplate(&b, baseFile, properties); err != nil {
+		return nil, err
+	}
+
+	err := json.Unmarshal(b.Bytes(), &paramsDescMap)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return paramsDescMap, nil
```

Comment:
Should we just return `paramsDescMap["parameters"]` here so we don't have to look it up later? Is there any other key in the unmarshalled map that is useful?

Reply:
yeah I can do that!